File: 1505-minimum-possible-integer-after-at-most-k-adjacent-swaps-on-digits/1505-minimum-possible-integer-after-at-most-k-adjacent-swaps-on-digits.cpp
class Fenwick {
public:
    explicit Fenwick(int n): sums_(n+1) {}
    void update(int i, int delta) {
        ++i;
        while(i < sums_.size()) {
            sums_[i] += delta;
            i += i & -i;
        }
    }
    int query(int i) {
        ++i;
        int ans = 0;
        while(i > 0) {
            ans += sums_[i];
            i -= i & -i;
        }
        return ans;
    }
  
private:
    vector<int> sums_;
};
class Solution {
public:
    string minInteger(string num, int k) {
        int n = num.size();
        vector<queue<int>> pos(10);
        for(int i = 0; i < n; ++i)
            pos[num[i]-'0'].push(i);
        Fenwick tree(n);
        vector<bool> used(n);
        string ans;
        
        while(ans.size() < n && k > 0) {
            for(int d = 0; d < 10; ++d) {
                if(pos[d].empty()) continue;
                int i = pos[d].front();
                int cost = i - tree.query(i-1);
                if(cost > k) continue;
                tree.update(i, 1);
                k -= cost;
                ans += d + '0';
                used[i] = true;
                pos[d].pop();
                break;
            }
        }
        for(int i = 0; i < n; ++i)
            if(!used[i]) ans += num[i];
        return ans;
      }
};

File: 1505-minimum-possible-integer-after-at-most-k-adjacent-swaps-on-digits/1505-minimum-possible-integer-after-at-most-k-adjacent-swaps-on-digits.java
class Solution {
  class Fenwick {
    private int[] sums;
    
    public Fenwick(int n) {
      this.sums = new int[n + 1];
    }
    
    public void update(int i, int delta) {
      ++i;
      while (i < sums.length) {
        this.sums[i] += delta;
        i += i & -i;
      }
    }
    
    public int query(int i) {
      int ans = 0;
      ++i;
      while (i > 0) {
        ans += this.sums[i];
        i -= i & -i;
      }
      return ans;
    }
  }
  
  public String minInteger(String num, int k) {
    int n = num.length();
    var used = new boolean[n];
    var pos = new ArrayList<ArrayDeque<Integer>>();
    for (int i = 0; i <= 9; ++i)
      pos.add(new ArrayDeque<Integer>());    
    for (int i = 0; i < num.length(); ++i)
      pos.get(num.charAt(i) - '0').offer(i);
    var tree = new Fenwick(n);
    var sb = new StringBuilder();
    while (k > 0 && sb.length() < n) {
      for (int d = 0; d <= 9; ++d) {
        Integer i = pos.get(d).peek();
        if (i == null) continue;
        int cost = i - tree.query(i - 1);
        if (cost > k) continue;        
        sb.append((char)(d + '0'));
        k -= cost;
        pos.get(d).removeFirst();
        tree.update(i, 1);
        used[i] = true;
        break;
      }
    }
    for (int i = 0; i < n; ++i)
      if (!used[i]) sb.append(num.charAt(i));
    return sb.toString();
  }
}